framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java

```diff
@@ -2615,10 +2616,27 @@ private void builderSetValue(AnnotationBuilder builder, String name, Object valu
    * @param annos the the set of declaration annotations on it, as written in the annotation file
    */
   private void putOrAddToDeclAnnos(String key, Set<AnnotationMirror> annos) {
-    if (annotationFileAnnos.declAnnos.containsKey(key)) {
-      annotationFileAnnos.declAnnos.get(key).addAll(annos);
-    } else {
+    Set<AnnotationMirror> stored = annotationFileAnnos.declAnnos.get(key);
+    if (stored == null) {
       annotationFileAnnos.declAnnos.put(key, new HashSet<>(annos));
+    } else {
+      if (fileType != AnnotationFileType.JDK_STUB) {
+        stored.addAll(annos);
+      } else {
+        // JDK annotations should not replace any annotation of the same type.
+        List<DeclaredType> storedAnnoTypes =
+            CollectionsPlume.mapList(AnnotationMirror::getAnnotationType, stored);
+        Types typeUtils = processingEnv.getTypeUtils();
+        for (AnnotationMirror anno : annos) {
+          DeclaredType annoType = anno.getAnnotationType();
+          boolean existing =
+              storedAnnoTypes.stream()
+                  .anyMatch(storedAnnoType -> typeUtils.isSameType(annoType, storedAnnoType));
+          if (!existing) {
```

Comment:
Instead of computing existing, you could just use `AnnotationUtils.containsSameByName(stored, anno)`.  That should be faster than comparing types.